### PR TITLE
Adding test for some HQL query scenarios

### PR DESCRIPTION
**Pull request by ammachado**
I'm trying to reuse the HQL parser as a generic JPQL parser and I've found that some of my scenarios are not able to be parsed using the current HqlParser.  Here is a PR with some test cases, both passing and failing test cases.

### CODE REVIEW

**Comment by sebersole** [comment]
In general terms, non-ORM uses of the parser are not at all a priority.  If
we can support such uses without compromising ORM's needs we will consider
it.

**Comment by ammachado** [comment]
Steve,

All cases reported are valid against JPA 2.1.

**Comment by sebersole** [comment]
Maybe.. I have not looked at the exact code.

But you said "I'm trying to reuse the HQL parser as a generic JPQL
parser"... I'm just pointing out that in general it is not our priority to
make this all work as a "generic parser".

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/sqm/produce/HqlParserTest.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+import org.junit.runners.Parameterized;
+
+import org.antlr.v4.runtime.BailErrorStrategy;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CodePointCharStream;
+import org.antlr.v4.runtime.CommonTokenStream;
+
+import static org.junit.Assert.fail;
+
+@RunWith(Parameterized.class)
```
Comment (sebersole) [inline hibernate-core/src/test/java/org/hibernate/orm/test/query/sqm/produce/HqlParserTest.java]: We use JUnit 5 on this branch.  JUnit 4 is available which is why this works, but the intention there is to simply allow existing tests to work for now.  All new tests should use JUnit 5.

Also, Parameterized is not a good paradigm here.  Simple tests like "does this query parse" are actually not useful any longer.  

And in fact, I'd be surprised if many (if not most) of these pass now (e.g. I just added support for update/delete queries late last week).

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/sqm/produce/HqlParserTest.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+				{ "SELECT bean.name FROM TestBean bean WHERE ((3 - 2) * 2) <= 10.0" },
+				{ "SELECT bean.name FROM TestBean bean WHERE (bean.name <> 'testBean')" },
+				{ "SELECT bean.name FROM TestBean bean WHERE bean.booleanValue = true" },
+				{ "SELECT bean1 FROM TestBean bean1, TestBean bean2 WHERE bean1.id < bean2.id" },
+				{ "SELECT bean1.name FROM TestBean bean1, TestBean bean2 WHERE (bean1.id - (bean2.id - 1)) = 5" },
+				{ "SELECT c FROM Customer c INNER JOIN c.orders o WHERE c.status = 1" },
+				{ "SELECT c FROM Customer c JOIN c.orders o WHERE c.status = 1" },
+				{ "SELECT c FROM Customer c WHERE (SELECT AVG(o.price) FROM c.orders o) > 100" },
+				{ "SELECT c FROM Customer c WHERE c.status = :stat" },
+				{ "SELECT c FROM Customer c WHERE FUNCTION(‘hasGoodCredit’, c.balance, c.creditLimit)" },
```
Comment (sebersole) [inline hibernate-core/src/test/java/org/hibernate/orm/test/query/sqm/produce/HqlParserTest.java]: Back-tick is not a valid quote character

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/sqm/produce/HqlParserTest.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+			fail( "failed to parse jpql:\n" + source + "\n\n" + message );
+		}
+	}
+
+	@Parameterized.Parameters(name = "{index}: {0}")
+	public static Iterable<Object[]> data() throws IOException {
+		return Arrays.asList( new Object[][] {
+				{ "DELETE FROM Customer c WHERE c.status = 'inactive'" },
+				{ "DELETE FROM Customer c WHERE c.status = 'inactive' AND c.orders IS EMPTY" },
+				{ "DELETE FROM Employee e" },
```
Comment (sebersole) [inline hibernate-core/src/test/java/org/hibernate/orm/test/query/sqm/produce/HqlParserTest.java]: DELETE seems to require WHERE atm which leads to more than one of 28 failures.  Which is strange though because I am pretty sure I am using that on the main 6.0 branch already

**Comment by sebersole** [comment]
The majority of "real" errors are related to things that are not even implemented yet:

1. sub-queries
2. (not) member-of
3. group-by/having
4. fetching

There were only 2 "bugs" that came from this:

1. DELETE requires a WHERE which causes a few failures.  This surprises me tbh.  I am fairly certain I have such queries in my latest 6.0 work and they work.
2. Grammar does not account for "grouping" an entire expression in parens.  E.g. `"... where (3+2) = 5"` fails to parse `(3+2)` due to the (completely unnecessary) parens around the expression

I'll investigate these last 2.

Anyway, I cannot integrate this as is.  The build is passing and this one causes it to fail.  Which is another reason this should not be using `@Parameterized` - we cannot apply `@FailureExpected`

**Comment by ammachado** [comment]
This last commit solves the grouping parens issue and all other grammar tests, except one for an edge case on `LIKE` expressions.

**Comment by sebersole** [comment]
Sweet.  I'll work on integrating these tomorrow

File: hibernate-core/src/main/antlr/org/hibernate/query/sqm/produce/internal/hql/grammar/HqlParser.g4 [hunk 4]
```diff
@@ -312,55 +310,58 @@ whereClause
...
+	| expression GREATER expression													# GreaterThanPredicate
+	| expression GREATER_EQUAL expression											# GreaterThanOrEqualPredicate
+	| expression LESS expression													# LessThanPredicate
+	| expression LESS_EQUAL expression												# LessThanOrEqualPredicate
+	| expression (NOT)? IN inList													# InPredicate
+	| expression (NOT)? BETWEEN expression AND expression							# BetweenPredicate
+	| expression (NOT)? LIKE expression (likeEscape)?								# LikePredicate
+	| MEMBER OF path																# MemberOfPredicate
+	| EXISTS LEFT_PAREN querySpec RIGHT_PAREN										# ExistsPredicate
 	;
```
Comment (sebersole) [inline hibernate-core/src/main/antlr/org/hibernate/query/sqm/produce/internal/hql/grammar/HqlParser.g4]: In the future, it would be nice if only real changes were included.  These are a lot of non-changes changes to sift through

File: hibernate-core/src/main/antlr/org/hibernate/query/sqm/produce/internal/hql/grammar/HqlParser.g4 [hunk 5]
```diff
@@ -771,3 +774,20 @@ identifier
...
+dotIdentifierOrParam
+    : dotIdentifierSequence
+    | parameter
+    ;
+
+dotIdentifierOrParamList
+    : dotIdentifierOrParam ( COMMA dotIdentifierOrParam )+
+    ;
+
+comparisonOperators
```
Comment (sebersole) [inline hibernate-core/src/main/antlr/org/hibernate/query/sqm/produce/internal/hql/grammar/HqlParser.g4]: This rule is never used?

Comment (ammachado) [reply hibernate-core/src/main/antlr/org/hibernate/query/sqm/produce/internal/hql/grammar/HqlParser.g4]: Those are never used.  I'll remove them.

File: hibernate-core/src/main/antlr/org/hibernate/query/sqm/produce/internal/hql/grammar/HqlParser.g4 (hunk 4 above)
Comment (ammachado) [reply hibernate-core/src/main/antlr/org/hibernate/query/sqm/produce/internal/hql/grammar/HqlParser.g4]: These changes were necessary to make test pass.

Comment (sebersole) [reply hibernate-core/src/main/antlr/org/hibernate/query/sqm/produce/internal/hql/grammar/HqlParser.g4]: Look at the diff..it is exactly the same.  Only formatting changed
